### PR TITLE
Add Docker Swarm standalone support

### PR DESCRIPTION
**Pull request by oeuftete**
### What does this PR do?

Updates the `dockerng` modules and states with support for [Docker Swarm standalone](https://github.com/docker/swarm).

### What issues does this PR fix or reference?

None.

### Tests written?

No.

----

FWIW, I've been using these same changes for several months against several `2015.8.10` production installations, some which use Docker Swarm standalone, some with just a traditional Docker engine.  We're preparing to move to `2016.11.x`, and I thought I should make an attempt to make these available to the community.  If they're too risky, that's OK.  :)

### CODE REVIEW

File: salt/states/dockerng.py [hunk 1]
```diff
@@ -421,7 +421,25 @@ def _image_get(path, default=None):
                     data = dict((k, '') for k in data)
                 image_labels.update(data)
                 data = image_labels
-            if actual_data != data:
+
+            def _filter_swarm_labels(_api_data):
```
Comment (oeuftete) [inline salt/states/dockerng.py]: Here's a bit more background.  The labels for a Docker Swarm standalone container could look something like this:

```json
{
  "com.docker.swarm.affinities": "[\"container!=~/myapp.*/\"]",
  "com.docker.swarm.id": "121b791058b8334c59d64b7ccbc5e73099e9b845cbf873bb69b48e6d32652fe5",
  "com.example.deployment_type": "development"
}
```

The idea is that `com.docker.swarm.id` doesn't need to be compared when seeing if a new container is required.

I honestly can't recall when the "empty" (`[]`) `com.docker.swarm.affinities` value could come up, but when I wrote this several months ago, I did need it.

**Comment by terminalmage** [comment]
First of all, feature additions should be made in the ``develop`` branch. Secondly, I have an open PR which is a day or two from being merged (https://github.com/saltstack/salt/pull/39996) which refactors our Docker support significantly in the develop branch.

I think we may need to hold off like something like this and find a way to integrate it with the work being done in develop once my PR is merged. One thing to keep in mind when reworking this code against the updated code in develop, is that ``dockerng`` is now simply ``docker`` (though ``dockerng`` can still be used in SLS files and on the CLI for the next couple releases).

**Comment by oeuftete** [comment]
I guess I can see this being treated as a feature.  I've sort of felt that since Swarm is supposed to be transparent to users of the Docker API, salt should just work.  But I can see your point.

I'll see if I can try out the refactored `develop` against our states sometime.  I'll close this and keep using my internal fork for our production needs.

**Comment by terminalmage** [comment]
@oeuftete After a 2nd and 3rd look, I can see how this would be construed as a bugfix, given that it would present a problem in comparing containers. I haven't personally used Docker Swarm, but it appears to me that it results in labels being added to the container by Docker, correct? In other words, these labels are not expected to be in the SLS file.

I'm going to reopen this for discussion purposes.

**Comment by terminalmage** [comment]
One thing I can tell you about this PR is that the change to the list_containers() function should apply cleanly when merged forward into develop, and work well in the develop branch (where its new home would be in salt/modules/docker.py). The code for the ``dockerng.running`` state however may need a little work to be supported in develop (and the code will live in ``salt/states/docker_container.py`` moving forward). To compare containers now, we create a new temporary container with the passed arguments, and then compare the Config and HostConfig in the two containers to see if there are differences in either. If there are, then it is assumed that we need to replace that container, and the old one is stopped and deleted, while the "temporary" container is then renamed and started, becoming the new container.

The important thing to note is when Docker would add those labels to the container. Does it happen at the time it is created, or not until it is started? If the former, then we would need to modify our comparison code to filter those Docker Swarm labels out of the diff.

**Comment by cachedout** [comment]
@oeuftete Did you see the latest feedback from @terminalmage above? Just checking in. :]

**Comment by oeuftete** [comment]
@cachedout @terminalmage Yep, saw it, might take me some time to get back to this, unfortunately.

**Comment by terminalmage** [comment]
@oeuftete I updated my comment from last week to be up-to-date (my PR is now merged, and we had to rename the docker execution module.

**Comment by cachedout** [comment]
@oeuftete Will you return to this one in the near future? If not, we'll likely close it and re-open it when you are are ready.

**Comment by oeuftete** [comment]
@cachedout Realistically, I probably won't get to it too soon.  Sorry. :(

**Comment by terminalmage** [comment]
@oeuftete I'm saving the diff here in case you close it, as I may take a crack at this if you don't have the time to do so. If I do work on this, I will be sure to loop you in.

**Comment by cachedout** [comment]
I'm going to go ahead and close this then. @oeuftete if you do decide to come back to this, just leave a comment and we'll be pleased to re-open it. Thanks!

**Pull request closed without merging**